accurate-gelu

jcrist1 opened the pull request:
This PR adds an accurate gelu function (used at least in in GPT2 model from huggingface) (see [Issue #804](https://github.com/coreylowman/dfdx/issues/804). Importantly in order to make the operators generic over  `Dtype`, I introduce an `Erf` trait that allows us to call `d_type.erf()` to get the error function of the value `d_type`. I currently am getting a compile error with feature `cuda`, and I have a hunch that this trait might be the issue. I'm having trouble debugging the build further as I don't have the cuda headers anywhere (working on a mac), so I would appreciate some help, so I was wondering if someone more familiar with the code might be able to point out the error.

For Github: Resolves #804

jcrist1 commented:
Also I'm not at sure that the cuda actually supports the erf as I can't build cuda locally.

jcrist1 commented:
Finally the naming is pretty terrible. Would it make sens to make a Gelu Activation an enum 
```rust
enum Gelu {
   Fast,
   Accurate
}
```
But then probably the `activation_impls!` macro couldn't be used

nkoppel commented:
Thank you for contributing! This looks good so far, I just have a few comments.
- To check the CUDA build without needing to call nvcc, use the command below. This won't be able to check your kernels, but should help you resolve your current CI errors.
```shell
cargo +nightly clippy -F cuda,ci-check
```
- I think it's best that we name the op/module 'AccurateGeLU' rather than 'GeLUCorrect' because it's easier to read and doesn't imply the the approximate GeLU op is incorrect.
- Be sure to document the difference between the accurate gelu op and the normal gelu.

jcrist1 commented on src/tensor_ops/accurate_gelu/cpu_kernel.rs:
```diff
@@ -0,0 +1,55 @@
+use libm::{erf, erff};
+use num_traits::{Float, FloatConst};
+
+trait Erf {
+    fn erf(self) -> Self;
+}
+
+#[cfg(feature = "f16")]
+impl Erf for f16 {
+    fn erf(self) -> Self {
+        f16::from_f32(erff(f16::to_f32(self)))
+    }
+}
+
+impl Erf for f64 {
+    fn erf(self) -> Self {
+        erf(self)
+    }
+}
+
+impl Erf for f32 {
+    fn erf(self) -> Self {
+        erff(self)
+    }
+}
```
I feel like these should be nearer to the Dtype trait, it feels a little odd putting them here, but I'm not that familiar with the code base so don't know the best place to put them yet.

coreylowman replied:
I actually like this location. Keeping things as close to where they are used is preferred IMO. Not sure if there's a standard practice for this though, but I like it 👍

jcrist1 commented:
Okay @nkoppel I updated the name. Accurate GeLU is much better. I think distinguishing that the other GeLU is faster will help. I couldn't find an explicit citation for the fact that it's faster, but it seems to only require a single exponential, while the error function requires a much higher degree polynomial and still an exponential. I also beefed up the docs.

I feel like the changes to the docs for the activations don't fit in the code, but I'm guessing most people using the code will be using them as activations, rather than postfix operations so I added some info there even though it breaks the nice code block.

coreylowman commented on src/tensor_ops/gelu/gelu.cu:
Are the changes in this file just indentation? Can you revert them if so? Just for easier reviewing 😀

jcrist1 replied:
should be, and will do. LSP automatically updated them.

coreylowman commented on src/tensor_ops/accurate_gelu/mod.rs:
````diff
@@ -0,0 +1,75 @@
+/// let t = dev.tensor([-1.0, 0.0, 1.0, 2.0]);
+/// let r = t.accurate_gelu();
+/// ```
+pub fn accurate_gelu<S: Shape, E: Dtype, D: UnaryKernel<AccurateGeLUKernelOp, E>, T: Tape<E, D>>(
````
Thoughts on calling this `true_gelu`? I wonder if we should rename the other one to `approx_gelu`.

nkoppel replied:
How should we deal with breaking name changes like this? A good approach may be to make aliases to the old naming marked as deprecated that direct users to the new names, at least for a major version or two.

coreylowman replied:
Yeah that'd be a good way forward if we do want to rename it.

jcrist1 replied:
I have to admit I like `fast` vs. `accurate`, because `true` and `approx` gelu doesn't really describe the tradeoff: if there is a "true" gelu why would you _not_ choose it? but if there is an accurate and a fast gelu, then I know how to make a decision. What do y'all think? Otherwise fine with the approach of using an alias.

jcrist1 replied:
I did it now with fast and accurate, but just to emphasize I'm not married to the naming. As repo owners tell me what's the best and I'll use that.

jcrist1 replied:
Didn't realise type aliased don't work for instantiating zero-size types, so I had to introduce a struct, but the implementation macro is not compatible with the `deprecated` macro, so I just implemented the traits by hand.

nkoppel commented:
Resolves #804.

nkoppel commented on src/tensor_ops/fast_gelu/mod.rs:
```diff
@@ -0,0 +1,79 @@
+}
+
+impl<S: Shape, E: Dtype, D: UnaryKernel<FastGeLUKernelOp, E>, T: Tape<E, D>> Tensor<S, E, D, T> {
+    /// See [gelu]
+    pub fn fast_gelu(self) -> Self {
+        self.try_fast_gelu().unwrap()
+    }
+    /// See [gelu]
+    pub fn try_fast_gelu(self) -> Result<Self, D::Err> {
+        try_unary_op(FastGeLUKernelOp, self)
+    }
+
+    #[deprecated(since = "0.12.0", note = "Use `fast_gelu` instead")]
+    pub fn gelu(self) -> Self {
+        self.fast_gelu()
+    }
+
+    #[deprecated(since = "0.12.0", note = "Use `try_fast_gelu` instead")]
+    pub fn try_gelu(self) -> Result<Self, D::Err> {
+        self.try_fast_gelu()
+    }
```
Top two methods should link to `fast_gelu`, and deprecated items should have a link to their non-deprecated counterparts.

nkoppel commented on src/tensor_ops/fast_gelu/mod.rs:
```diff
@@ -0,0 +1,79 @@
+#[derive(Debug, Default, Copy, Clone)]
+pub struct FastGeLUKernelOp;
+
+/// [Fast Gaussian Linear Unit (GeLU)](https://paperswithcode.com/method/gelu). `0.5 * x * (1 + tanh(sqrt(2 / pi) * (x + 0.044715 * x^3)))`
```
These docs should include a link to AccurateGeLU

nkoppel commented on src/tensor_ops/accurate_gelu/mod.rs:
````diff
@@ -0,0 +1,75 @@
+/// GeLU(x) ~ 0.5 ∗ x ∗ (1.0 + tanh((sqrt(2.0/π) ∗ (x + 0.044715 ∗ x^3)))
+/// ```
+///
+/// See [gelu](crate::tensor_ops::gelu::gelu) to use this approximation
````
This link needs to be fixed with the new naming

nkoppel commented on src/nn/activations.rs:
```diff
@@ -38,6 +43,25 @@ activation_impls!(Abs, try_abs, #[doc="Calls [abs()]."]);
 activation_impls!(Softmax, try_softmax, #[doc="Calls [softmax()]."]);
 activation_impls!(LogSoftmax, try_log_softmax, #[doc="Calls [log_softmax()]."]);
 
+#[deprecated(since = "0.12.0", note = "please use `FastGeLU` instead")]
+#[derive(Default, Debug, Clone, Copy)]
+pub struct GeLU;
```
Needs to link to it's non-deprecated counterpart.

jcrist1 replied:
link is 3 lines up

nkoppel reviewed (commented):
Just a few little fixes to documentation and this should be good to go! 

Note that I don't have write access to dfdx, I've just contributed a lot, so @coreylowman gets the final say on everything.

jcrist1 commented:
@nkoppel alright should be addressed

coreylowman commented:
This looks good from me - any other updates planned here?

jcrist1 commented:
I'm not planning on anything new.

coreylowman commented:
We can open another PR if there's something else to add/change!
